### PR TITLE
Working with CSV test files - issues #831, #832 , #833 and #840

### PR DESCRIPTION
**Pull request by fjabeen02**
# Terms

> THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE TERMS OF THE CCLA DATED 2017-11-07 WITH FINOS/LINUX FOUNDATION (FORMERLY THE SYMPHONY SOFTWARE FOUNDATION CCLA).
>
> THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

File: tests-integration/spark/test/src/SparkTypeTest.scala [hunk 1]
```diff
@@ -10,153 +11,242 @@ class SparkTypeTest extends FunSuite {
...
+    df_actual_results.show(false)
+    assert(df_actual_results.columns.length == df_expected_results.columns.length)
+    assert (df_actual_results.count == df_expected_results.count)
+    assert(df_actual_results.except(df_expected_results).isEmpty && df_expected_results.except(df_actual_results).isEmpty)
+  }
+
+  test("testMaybeString") {
+    val df_expected_results = localTestSession.read.format("csv")
+      .option("header", "true")
+      .schema(foo_string_schema)
```
[the remark this answers is not in the document]
Comment (jonathanmaw) [reply tests-integration/spark/test/src/SparkTypeTest.scala:150]: Likewise for all the other tests of maybes

**Review by jonathanmaw — CHANGES_REQUESTED** [review]
This all looks good (though perhaps I've had too much of a hand in it to have perspective about the general shape of it), and a lot of hard work has *definitely* gone into it.

In summary:
* There are a few very minor formatting errors (strange newlines, incorrect capitalisation, weird typos)
* I've had some thoughts on possible improvements to the csv file that might be worth making a task of to do later.
* There are a bunch of types and functions in CsvUtils that I'd like to know if we can do without them.

I also couldn't make an inline comment about it, but "tests-integration/spark/test/src/spark_test_data/expected_results_TestAntiqueSSWhere1.csv" doesn't appear to be used anywhere and can be deleted.

File: package-lock.json [hunk 2]
```diff
@@ -5,19 +5,21 @@
   "requires": true,
   "packages": {
     "": {
-      "name": "morphir-elm",
```
[the remark this answers is not in the document]
Comment (fjabeen02) [reply package-lock.json:8]: Yeah, when I didnt commit the changes it caused that weird npm error I had a couple days ago

File: package.json [hunk 3]
```diff
@@ -65,7 +65,7 @@
     "@vercel/ncc": "^0.24.1",
     "del-cli": "3.0.1",
     "elm": "^0.19.1-5",
-    "elm-doc-preview": "^3.0.4",
+    "elm-doc-preview": "^1.0.7",
```
[the remark this answers is not in the document]
Comment (fjabeen02) [reply package.json:68]: I didnt use them but I had to upgrade them to fix that straneg npm error I got a couple days ago.

**Comment by jonathanmaw** [comment]
I think you accidentally committed a log file in `tests-integration/out.txt`

**Review by jonathanmaw — APPROVED** [review]

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (AttilaMihaly) [reply package.json:68]: I just realized that you actually downgraded `elm-doc-preview`. Is there a reason for that?

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]
I need to see a documentation on how this testing framework works. I tried to infer it based on the files being added but it's very difficult to see what things are for.

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (fjabeen02) [reply package.json]: I think I must have done that whilst trying to debug a version error I was getting. I have reverted back to 3.0.4 now

**Comment by fjabeen02** [comment]
@AttilaMihaly  I have now uploaded the documentation to this PR

**Comment by AttilaMihaly** [comment]
> @AttilaMihaly I have now uploaded the documentation to this PR

Thanks. While I review, can you resolve the conflicts?

**Review by AttilaMihaly — CHANGES_REQUESTED** [review]
It looks good apart from the minor issues I highlighted in the doc.

File: docs/developers-guide/files/spark-testing-framework.md [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                    (\example ->
+                        [ ( "name", example.name )
+                        , ( "length", String.fromFloat example.length )
+                        ]
+                    )
+            , fieldSeparator = ','
+            }
+```
+
+#### 4) Write the code that will run the test on the input data and produce an output data csv
```
[the remark this answers is not in the document]
Comment (fjabeen02) [reply docs/developers-guide/files/spark-testing-framework.md:148]: I have fixed both of these mistakes now :)

**Review by AttilaMihaly — APPROVED** [review]

**Pull request merged**